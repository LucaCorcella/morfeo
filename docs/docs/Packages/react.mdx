---
sidebar_position: 3
id: react
title: react
description: morfeo's react package
---

## Installation

```bash
# npm
npm i @morfeo/react

# yarn
yarn add @morfeo/react
```

**@morfeo/react** re-expose [@morfeo/hooks](./hooks.mdx) and [@morfeo/web](./web.mdx), we suggest you to read their documentations.
