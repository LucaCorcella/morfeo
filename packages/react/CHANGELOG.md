# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.2.0 (2021-08-08)


### Features

* documentation improvements ([4ad3b5d](https://github.com/VLK-STUDIO/morfeo/tree/main/packages/react/commit/4ad3b5d7f35cd9c1ad1532e5367dec21594d8ff4))
* website link replaced inside readme files ([1e55a6d](https://github.com/VLK-STUDIO/morfeo/tree/main/packages/react/commit/1e55a6d458d2873d09efd5fad5100cbbae382057))
* **react:** introducing react package ([f99da40](https://github.com/VLK-STUDIO/morfeo/tree/main/packages/react/commit/f99da40a8b8fe0c440002e1065dbc73f7376f732))





## 0.1.8 (2021-07-19)


### Features

* documentation improvements ([4ad3b5d](https://github.com/VLK-STUDIO/morfeo/tree/main/packages/react/commit/4ad3b5d7f35cd9c1ad1532e5367dec21594d8ff4))
* website link replaced inside readme files ([1e55a6d](https://github.com/VLK-STUDIO/morfeo/tree/main/packages/react/commit/1e55a6d458d2873d09efd5fad5100cbbae382057))
* **react:** introducing react package ([f99da40](https://github.com/VLK-STUDIO/morfeo/tree/main/packages/react/commit/f99da40a8b8fe0c440002e1065dbc73f7376f732))





## 0.1.7 (2021-07-14)


### Features

* documentation improvements ([4ad3b5d](https://github.com/VLK-STUDIO/morfeo/tree/main/packages/react/commit/4ad3b5d7f35cd9c1ad1532e5367dec21594d8ff4))
* website link replaced inside readme files ([1e55a6d](https://github.com/VLK-STUDIO/morfeo/tree/main/packages/react/commit/1e55a6d458d2873d09efd5fad5100cbbae382057))
* **react:** introducing react package ([f99da40](https://github.com/VLK-STUDIO/morfeo/tree/main/packages/react/commit/f99da40a8b8fe0c440002e1065dbc73f7376f732))





## 0.1.6 (2021-06-27)


### Features

* website link replaced inside readme files ([1e55a6d](https://github.com/VLK-STUDIO/morfeo/tree/main/packages/react/commit/1e55a6d458d2873d09efd5fad5100cbbae382057))
* **react:** introducing react package ([f99da40](https://github.com/VLK-STUDIO/morfeo/tree/main/packages/react/commit/f99da40a8b8fe0c440002e1065dbc73f7376f732))





## 0.1.5 (2021-05-23)


### Features

* **react:** introducing react package ([f99da40](https://github.com/VLK-STUDIO/morfeo/tree/main/packages/web/commit/f99da40a8b8fe0c440002e1065dbc73f7376f732))
