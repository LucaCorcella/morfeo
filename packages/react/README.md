<div align="center">
<h1>@morfeo/react</h1>
</div>

**@morfeo/react** is part of the [@morfeo](https://github.com/VLK-STUDIO/morfeo) eco-system, a set of **framework-agnostic** tools that help you to create beautiful design systems for your web and mobile apps.

---

<div align="center">
  <a href="https://github.com/VLK-STUDIO/morfeo">Documentation</a> |
  <a href="https://github.com/VLK-STUDIO/morfeo">API</a> |
  <a href="https://github.com/VLK-STUDIO/morfeo">Contributing</a> |
  <a href="https://morfeo.slack.com">Slack</a>
</div>

---
