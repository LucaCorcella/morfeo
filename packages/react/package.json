{
  "name": "@morfeo/react",
  "author": {
    "name": "Mauro Erta",
    "email": "mauro@vlkstudio.com"
  },
  "private": false,
  "version": "0.2.0",
  "license": "MIT",
  "main": "build/index.js",
  "module": "build/index.js",
  "types": "build/index",
  "typings": "build/index",
  "keywords": [
    "design",
    "system",
    "morfeo",
    "morfeo-js",
    "react"
  ],
  "scripts": {
    "build": "rimraf build && tsc",
    "watch": "tsc -w"
  },
  "dependencies": {
    "@morfeo/hooks": "^0.2.0",
    "@morfeo/web": "^0.2.0",
    "fast-safe-stringify": "^2.0.8"
  },
  "peerDependencies": {
    "csstype": "^3.0.8",
    "react": "^17.0.2",
    "react-dom": "^17.0.2"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "build"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/VLK-STUDIO/morfeo/tree/main/packages/react"
  },
  "homepage": "https://morfeo.dev",
  "bugs": {
    "url": "https://github.com/VLK-STUDIO/morfeo/issues"
  }
}
