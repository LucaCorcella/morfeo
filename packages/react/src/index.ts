export * from '@morfeo/web';
export * from '@morfeo/hooks';
export * from './useClassName';
export * from './MorfeoComponent';

export { useStyle, useStyles } from './hooks';
