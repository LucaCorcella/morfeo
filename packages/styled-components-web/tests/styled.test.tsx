import React from 'react';
import { theme } from '@morfeo/react';
import morfeoStyled from '../src';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

beforeAll(() => {
  theme.set({
    colors: {
      primary: 'black',
      secondary: 'white',
    },
    components: {
      Box: {
        tag: 'div',
        style: { color: 'primary' },
        props: { 'aria-label': 'default box' },
        variants: {
          primary: {
            tag: 'button',
            style: { color: 'secondary' },
            props: { 'aria-label': 'primary button', type: 'submit' },
          },
          secondary: {
            style: { bg: 'secondary' },
            props: { 'aria-label': 'secondary button', type: 'submit' },
          },
        },
      },
      Text: {
        tag: 'p',
        style: {},
        props: { variant: 'h1' },
        variants: {
          h1: {
            tag: 'h1',
            style: { fontWeight: 'bold' },
          },
        },
      },
      Incomplete: {},
    },
  } as any);
});

describe('morfeoStyled', () => {
  test('should work as the default styled interface', () => {
    const Button = morfeoStyled.button`
      color: red;
    `;
    const tree = renderer.create(<Button />).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveStyleRule('color', 'red');
  });

  test('should resolve values from the theme if CSSObject is passed', () => {
    const Box = morfeoStyled.Box({});
    const tree = renderer.create(<Box>Test</Box>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should return a component even if default style is not passed', () => {
    // @ts-ignore
    const Box = morfeoStyled.Box();
    const tree = renderer.create(<Box>Test</Box>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should change component tag if described inside variant', () => {
    const Box = morfeoStyled.Box({ variant: 'primary' });
    const tree = renderer.create(<Box variant="primary">Test</Box>).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveProperty('type', 'button');
  });

  test('should have the component tag of the parent component', () => {
    const Box = morfeoStyled.Box({});
    const tree = renderer
      .create(<Box variant={'secondary' as any}>Test</Box>)
      .toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveProperty('type', 'div');
  });

  test('should set the default variant and tag based on theme props', () => {
    // @ts-ignore
    const Text = morfeoStyled.Text({});
    const tree = renderer.create(<Text>Test</Text>).toJSON();
    expect(tree).toMatchSnapshot();
    expect(tree).toHaveProperty('type', 'h1');
  });

  test('should not be a component if the tag is invalid', () => {
    const InvalidComponent = morfeoStyled.prototype;
    expect(typeof InvalidComponent).not.toBe('function');
  });

  test('should render a div if the component is missing information about the tag', () => {
    // @ts-ignore
    const Incomplete = morfeoStyled.Incomplete({});
    const tree = renderer.create(<Incomplete />).toJSON();
    expect(tree).toHaveProperty('type', 'div');
  });
});
