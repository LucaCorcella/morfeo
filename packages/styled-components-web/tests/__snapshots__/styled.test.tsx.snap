// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`morfeoStyled - when morfeo was not set should render the component 1`] = `
<div
  className="div-jDhEeQ"
>
  Test
</div>
`;

exports[`morfeoStyled should change component tag if described inside variant 1`] = `
.c0 {
  color: white;
}

<button
  aria-label="primary button"
  className="c0"
  type="submit"
>
  Test
</button>
`;

exports[`morfeoStyled should have the component tag of the parent component 1`] = `
.c0 {
  color: black;
  background-color: white;
}

<div
  aria-label="secondary button"
  className="c0"
  type="submit"
>
  Test
</div>
`;

exports[`morfeoStyled should resolve values from the theme if CSSObject is passed 1`] = `
.c0 {
  color: black;
}

<div
  aria-label="default box"
  className="c0"
>
  Test
</div>
`;

exports[`morfeoStyled should return a component even if default style is not passed 1`] = `
.c0 {
  color: black;
}

<div
  aria-label="default box"
  className="c0"
>
  Test
</div>
`;

exports[`morfeoStyled should set the default variant and tag based on theme props 1`] = `
.c0 {
  font-weight: bold;
}

<h1
  className="c0"
>
  Test
</h1>
`;

exports[`morfeoStyled should work as the default styled interface 1`] = `
.c0 {
  color: red;
}

<button
  className="c0"
/>
`;
